compiler/optimizations/inferConstRefs.cpp

```diff
@@ -0,0 +1,377 @@
+/*
+ * Copyright 2004-2016 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "optimizations.h"
+
+#include "astutil.h"
+#include "expr.h"
+#include "stmt.h"
+#include "stlUtil.h"
+#include <deque>
+
+class RefInfo {
+  public:
+    bool                 finalized; // Indicates if analysis is done
+    Symbol*              sym;
+    std::deque<SymExpr*> todo;
+    RefInfo(Symbol*);
+};
+
+RefInfo::RefInfo(Symbol* s) {
+  sym = s;
+  finalized = sym->qualType().getQual() == QUAL_CONST_REF;
+}
+
+std::map<Symbol*, RefInfo*> infoMap;
+typedef std::map<Symbol*, RefInfo*>::iterator RefInfoIter;
+
+static bool inferConstRef(Symbol*);
+
+static bool handleRHS(CallExpr* parent, CallExpr* call, SymExpr* se) {
+  SymExpr* LHS = toSymExpr(parent->get(1));
+  switch (call->primitive->tag) {
+    case PRIM_GET_MEMBER:
+    case PRIM_GET_MEMBER_VALUE:
```

Comment:
I added them because they can return references if the field is a reference. The case where 'use' is the base aggregate type and the field is not a reference is handled by the ref-ness check on LHS.